src/dmd/expressionsem.d

```diff
@@ -6383,6 +6397,42 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             return setError();
         }
 
+        if (tob.ty == Tarray && t1b.ty == Tarray && exp.e1.op != TOK.string_)
+        {
+            auto tFrom = t1b.nextOf();
+            auto tTo = tob.nextOf();
+
+            uint fromSize = cast(uint)tFrom.size();
+            uint toSize = cast(uint)tTo.size();
+
+            if (fromSize != toSize)
+            {
+                // Array element sizes do not match, so we must adjust the dimensions
+
+                if ((fromSize % toSize) != 0)
+                {
+                    // Runtime check needed in case arrays don't line up
+
+                    // lower to `object.__ArrayCast!(TFrom, TTo)(from)`
+
+                    auto id = Identifier.idPool("__ArrayCast");
```

Comment:
How can I fully qualify this to `object.__ArrayCast`?

Reply:
Look at how the comparison lowering is done to `__cmp`.

https://github.com/dlang/dmd/blob/03e7693150c453055b098768e59ebd31cb3ca52e/src/dmd/expressionsem.d#L9794-L9798

Reply:
But I need an `Identifer`, not an `Expression`, for the `TemplateInstance`, later.  How do I create a fully-qualified `Identifier`?

Reply:
> How do I create a fully-qualified Identifier?

I don't think that is possible. To get the fully qualified name inside the compiler one needs to walk up the parent chain and build up the name that way. In your case, I think you need to use the overload of the `TemplateInstance` constructor that takes a `TemplateDeclaration`. Perhaps you can store the `TemplateDeclaration` when it's analyzed in the same way as the `Object` class declaration (and others) is stored:

https://github.com/dlang/dmd/blob/03e7693150c453055b098768e59ebd31cb3ca52e/src/dmd/dclass.d#L367-L372